kv/dist_sender.go

```diff
@@ -491,198 +608,138 @@ func (ds *DistSender) sendRPC(raftID int64, replicas replicaSlice,
-			// return retriable errors if everything is unreachable. That seems
-			// awkward; rpcError should probably not be retriable (rpc.Send
-			// will still succeed if it manages to talk to *anyone*), and
-			// then we can clear the cache here only for retriable errors.
-			// It amounts to the same situations though.
-			if rpcErr != nil {
-				ds.updateLeaderCache(proto.RaftID(desc.RaftID), proto.Replica{})
-				ds.rangeCache.EvictCachedRangeDescriptor(args.Header().Key, desc)
-			}
-
-			if err != nil {
-				if log.V(1) {
-					log.Warningf("failed to invoke %s: %s", call.Method(), err)
-				}
+	// Retry logic for lookup of range by key and RPCs to range replicas.
+	retryOpts := ds.rpcRetryOptions
+	retryOpts.Tag = "routing " + call.Method().String() + " rpc"
 
-				// If retryable, allow retry. For range not found or range
-				// key mismatch errors, we don't backoff on the retry,
-				// but reset the backoff loop so we can retry immediately.
-				switch err.(type) {
-				case *proto.RangeNotFoundError, *proto.RangeKeyMismatchError:
-					// Range descriptor might be out of date - evict it.
-					ds.rangeCache.EvictCachedRangeDescriptor(args.Header().Key, desc)
-					// On addressing errors, don't backoff; retry immediately.
-					return retry.Reset, nil
-				case *proto.NotLeaderError:
-					leader := err.(*proto.NotLeaderError).GetLeader()
-					if leader != nil {
-						// First, verify that leader is a known replica; if not,
-						// we've got a stale replica; evict cache and store leader.
-						if i := replicas.FindReplica(leader.StoreID); i == -1 {
-							ds.rangeCache.EvictCachedRangeDescriptor(args.Header().Key, desc)
-						}
-						ds.updateLeaderCache(proto.RaftID(desc.RaftID), *leader)
-					}
-					return retry.Reset, nil
-				default:
-					if retryErr, ok := err.(util.Retryable); ok && retryErr.CanRetry() {
-						return retry.Continue, nil
-					}
-				}
-				return retry.Break, err
-			}
+	curReply := finalReply
+	for {
+		call.Reply = curReply
+		curReply.Header().Reset()
 
-			// If this request has a bound, such as MaxResults in
-			// ScanRequest, check whether enough rows have been retrieved.
-			if args, ok := args.(proto.Bounded); ok && args.GetBound() > 0 {
-				bound := args.GetBound()
-				if reply, ok := reply.(proto.Countable); ok {
-					if nextBound := bound - reply.Count(); nextBound > 0 {
-						// Update bound for the next round.
-						args.SetBound(nextBound)
-					} else {
-						// Set flag to break the loop.
-						descNext = nil
-					}
-				}
+		var desc, descNext *proto.RangeDescriptor
+		err := retry.WithBackoff(retryOpts, func() (retry.Status, error) {
+			var descErr error
+			// Get range descriptor (or, when spanning range, descriptors).
+			// sendAttempt below may clear them on certain errors, so we
+			// refresh (likely from the cache) on every retry.
+			desc, descNext, descErr = ds.getDescriptors(call)
+			// If getDescriptors fails, we don't fish for retryable errors.
```

Comment:
@spencerkimball I think those were retried before, but it seems that any range lookups that could reasonably fail would call Send() themselves and then retry within, correct? Getting something retryable back here would correspond to RPC errors or client retry errors from the lookup (which shouldn't occur).